File: src/bin/0417-pacific-atlantic-water-flow.py
# @leetup=info id=417 lang=python3 slug=pacific-atlantic-water-flow



# @leetup=code

from collections import deque


class Solution:
    def pacificAtlantic(self, heights: list[list[int]]) -> list[list[int]]:
        def bfs(q: deque) -> set:
            visited = set()
            while q:
                (r, c) = q.popleft()
                if (r, c) in visited:
                    continue
                visited.add((r, c))

                for dr, dc in ((0, 1), (0, -1), (-1, 0), (1, 0)):
                    nr = r + dr
                    nc = c + dc
                    if 0 <= nr < R and 0 <= nc < C and heights[r][c] <= heights[nr][nc]:
                        q.append((nr, nc))
            return visited

        R = len(heights)
        C = len(heights[0])

        q = deque()
        for r in range(R):
            q.append((r, 0))
        for c in range(1, C):
            q.append((0, c))

        visited1 = bfs(q)

        q = deque()
        for r in range(R):
            q.append((r, C - 1))
        for c in range(C - 1):
            q.append((R - 1, c))

        visited2 = bfs(q)

        return [list(a) for a in visited1.intersection(visited2)]


# @leetup=code

File: src/bin/0417-pacific-atlantic-water-flow.rs
// @leetup=info id=417 lang=rust slug=pacific-atlantic-water-flow

// @leetup=code

use std::collections::*;

impl Solution {
    pub fn pacific_atlantic(heights: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let r_len = heights.len();
        let c_len = heights[0].len();
        let solve = Solve {
            heights,
            r_len,
            c_len,
        };
        solve.solve()
    }
}

struct Solve {
    heights: Vec<Vec<i32>>,
    r_len: usize,
    c_len: usize,
}

impl Solve {
    fn solve(&self) -> Vec<Vec<i32>> {
        let mut q = VecDeque::new();
        for r in 0..self.r_len {
            q.push_back((r, 0));
        }
        for c in 0..self.c_len {
            q.push_back((0, c));
        }
        let visited1 = self.bfs(q);

        let mut q = VecDeque::new();
        for r in 0..self.r_len {
            q.push_back((r, self.c_len - 1));
        }
        for c in 0..self.c_len {
            q.push_back((self.r_len - 1, c));
        }
        let visited2 = self.bfs(q);

        visited1
            .intersection(&visited2)
            .into_iter()
            .copied()
            .map(|(r, c)| vec![r as i32, c as i32])
            .collect()
    }

    fn bfs(&self, mut q: VecDeque<(usize, usize)>) -> HashSet<(usize, usize)> {
        let mut visited = HashSet::new();
        while let Some((r, c)) = q.pop_front() {
            if !visited.insert((r, c)) {
                continue;
            }
            for (dr, dc) in [(0, 1), (0, !0), (1, 0), (!0, 0)] {
                let (nr, nc) = (r.wrapping_add(dr), c.wrapping_add(dc));
                if nr < self.r_len && nc < self.c_len && self.heights[r][c] <= self.heights[nr][nc]
                {
                    q.push_back((nr, nc));
                }
            }
        }
        visited
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::pacific_atlantic;
}
